### DISCUSSION TITLE
Possible to script jukebox?

### DISCUSSION THREAD

**Discussion by aereaux** [jcorporation/myMPD#807] [Q&A] [answered]
Hello,

I've been using mympd for a while and it works great.  I have some scripts I use to listen to random songs that mostly go off sets of playlists instead of individual songs so that I listen to certain tracks (like classical music) in order.  Is there some way to script the included jukebox so that I can include this behavior?

Also, looking a bit at the code, are there only a couple supported values of `jukeboxUniqueTag`, or can I set it to basically whatever I want?

Thanks.

**Comment by jcorporation** [comment, marked as answer]
The jukebox is fully scriptable as every function of myMPD. Look at the jsonrpc api documentation.

https://jcorporation.github.io/myMPD/references/api/

You can use the integrated lua scripting or call the api directly with e.g. curl.

The uniq tags of are not limited, but there is only one tag supported.

**Reply by aereaux** [reply]
Looking at the API I don't see a way to control the jukebox queue, I only see methods to clear/rm/list it.  Is there some way I can script which items get added to the queue?

**Reply by jcorporation** [reply]
You search for `MYMPD_API_PLAYER_OPTIONS_SET`

**Reply by aereaux** [reply]
It looks like that is pretty limited in terms of what behavior I can change.  I was hoping I could change it basically to choose a playlist from a certain directory to add.  My (shell) script right now is basically:
```
find "$DIR" -type f -name '*.m3u' -printf "%P\n" | shuf -n $1 | mpc load
```

**Reply by jcorporation** [reply]
Right, the jukebox could not select a random playlists. It selects random songs from a playlist, but you can script this with lua inside myMPD.

This selects a random playlists and replaces the queue and starts playing.

```
-- get the first 2000 playlists
rc, raw_result = mympd_api_raw("MYMPD_API_PLAYLIST_LIST", json.encode({offset = 0, limit = 2000, searchstr = "", type = 0}))
result = json.decode(raw_result)
-- random number
math.randomseed(os.time())
number = math.random(0, #result.result.data)
-- get playlist by random number
playlist = result.result.data[number].uri
-- play the playlist
rc, raw_result = mympd_api_raw("MYMPD_API_QUEUE_REPLACE_PLAYLIST", json.encode({plist = playlist, play = true}))
-- return the playlist name
return "Playing" .. playlist
```

**Reply by aereaux** [reply]
Thanks for the quick responses.  I've added grouping tags to my library, and now I'm trying to use that.  However, when I try to list the groupings I don't get anything (and same result using the API in a script).  Any idea what's wrong?
```
$ curl -k https://localhost:4430/api/ --data '{"jsonrpc":"2.0","id":0,"method":"MYMPD_API_DATABASE_TAG_LIST","params":{"offset":0,"limit":2000,"searchstr":"((Album == \'Yellow House\'))","sortdesc":false,"tag":"Grouping"}}'
{"jsonrpc":"2.0","id":0,"result":{"method":"MYMPD_API_DATABASE_TAG_LIST","data":[],"totalEntities":0,"returnedEntities":0,"offset":0,"searchstr":"((Album == 'Yellow House'))","tag":"Grouping","pics":false}}
```
```
$ mpc list Grouping "((Album == 'Yellow House'))"

Colorado
Knife
Little Brother
On a Neck, on a Spit
Plans
```

**Reply by jcorporation** [reply]
MYMPD_API_DATABASE_TAG_LIST does not support mpd filter expressions. `searchstr` is only matched against the tag values of `Grouping`. In the moment there is no API method to get the same results as with mpc.

**Reply by aereaux** [reply]
Ah, that makes sense, thanks!
